src/util/compilationDatabase.ml

```diff
@@ -61,24 +61,30 @@ let load_and_preprocess ~all_cppflags filename =
           (* TODO: extract o_file *)
           let command = reroot command in
           let preprocess_command = Str.replace_first command_program_regexp ("\\1 " ^ String.join " " (List.map Filename.quote all_cppflags) ^ " -E -MMD -MT " ^ file) command in
-          let preprocess_command = Str.replace_first command_o_regexp ("-o " ^ preprocessed_file) preprocess_command in
           if preprocess_command = command then (* easier way to check if match was found (and replaced) *)
-            failwith "CompilationDatabase.preprocess: no -o argument found for " ^ file
+            failwith "CompilationDatabase.preprocess: no program found for " ^ file
           else
-            preprocess_command
+            let preprocess_command_o = Str.replace_first command_o_regexp ("-o " ^ preprocessed_file) preprocess_command in
+            if preprocess_command_o = preprocess_command then (* easier way to check if match was found (and replaced) *)
+              preprocess_command ^ " -o " ^ preprocessed_file
+            else
+              preprocess_command_o
         | None, Some arguments ->
           let arguments = List.map reroot arguments in
-          begin match List.findi (fun i e -> e = "-o") arguments with
-            | (o_i, _) ->
-              begin match List.split_at o_i arguments with
-                | (arguments_program :: arguments_init, _ :: o_file :: arguments_tl) ->
-                  let preprocess_arguments = all_cppflags @ "-E" :: "-MMD" :: "-MT" :: file :: arguments_init @ "-o" :: preprocessed_file :: arguments_tl in
-                  Filename.quote_command arguments_program preprocess_arguments
-                | _ ->
-                  failwith "CompilationDatabase.preprocess: no -o argument value found for " ^ file
-              end
-            | exception Not_found ->
-              failwith "CompilationDatabase.preprocess: no -o argument found for " ^ file
+          let preprocess_arguments =
+            let suf =
+              begin match List.findi (fun i e -> e = "-o") arguments with
+                | (o_i, _) ->
+                  begin match List.split_at o_i arguments with
+                    | (arguments_init, o_file :: arguments_tl) -> arguments_init @ preprocessed_file :: arguments_tl
```

Comment:
Previously the pattern contained `_ :: o_file`, where the wildcard was used to skip the `-o` argument. Is this somehow not necessary now?

Reply:
It still is. The string concatenation was mixed up, but I hope I fixed it now. Can you take another look at it?